### ISSUE TITLE
Passwordless Authentication

### ISSUE THREAD

**Issue by thejoefriel** [yalla-coop/earwig#20]
Hi @theharrisonmoore - we've been reviewing the authentication strategy with @Jbarget and we think the best way to proceed would be to use auth0 and implement a passwordless strategy : https://auth0.com

What this does is allow users to sign up and log in without them needing to create or for us to store a password - they can simply use their email. This will then send them a unique code via email which they then enter to sign up / log in. 

Additionally, rather than needing to request 'remember me', best practice now is to assume they want to stay logged in, but that this would automatically expire after a certain period if they did not visit the site. It is also protected against brute-force log in attempts which is great extra level of security. 

So this kind of set up makes it super easy for you to onboard users, and is as safe and secure as implementing a password. 

Auth0 lets you have up to 7,000 free active users and unlimited logins so is absolutely perfect for start ups like yours.

Are you happy for us to proceed down this route? Let me know if you have any questions at all.

**Comment by theharrisonmoore** [comment]
Hi,

It sounds fine in principle. Can I check:

   1. How does a user log in once they've been timed out? Would they then
   have to search for the historic email? Or is there an easier way?
   2. What happens once we get close to 7,000 signups? How much development
   time for one dev would it take to then implement a different strategy?

<https://signup.earwig.work/>

**Comment by thejoefriel** [comment]
If they've been timed out, when they go to log in it would again send a brand new unique code to their email for them to then type in.

If you were going to go over the 7,000 sign ups you could either look at the pricing model to continue scaling with auth0 in which case nothing would change, or in terms of development time I'd say it would be a few days work to enable a different strategy - @Jbarget, what do you think in terms of the complexity of changing it if earwig got beyond 7,000 users?

**Comment by theharrisonmoore** [comment]
So it's not a link in the email they can click, but a code they must
copy/memorise and then paste/type into a field?

<https://signup.earwig.work/>

**Comment by thejoefriel** [comment]
It can be a link @theharrisonmoore 
This would be a 'magic link' which clicking it would then log them in

**Comment by theharrisonmoore** [comment]
That seems simple enough.

And, given that my goal for 2019 is to get 10,000 registered users, we
ought to assume earwig will exceed the 7,000 user count before too long.
Could you tell me what the Auth0 costs would be for the following:

   - 10,000 users
   - 50,000 users
   - 250,000 users

<https://signup.earwig.work/>

**Comment by Jbarget** [comment]
Building a secure authentication strategy would take at least a week I imagine
Pro's of auth0:
- quicker than building custom solution
- good documentation/instructions
- sensitive information isnt stored in our database

Cons:
- potential cost after 7000 users
- Auth0 API could change leading to necessary changes on our end (but they usually change for a good reason if they do)

In terms of the "unique code" vs. "magic link" a link that the user clicks and opens another window with the user logged in:

Because we're using React which is a one page app, if we load the webpage to show the login screen and then a user clicks on the magic link, it means that the whole app will load again in a new window. Whereas if we use a unique code the webpage is already loaded and the user enters the code on the same page. (i find it annoying when a website opens up new windows when its not expected)

**Comment by theharrisonmoore** [comment]
Thanks for the clarification Justen.

In that case, unless I've misunderstood, I have to say that the Auth0
approach seems to add more friction than asking a user to create a
password. Correct me if I'm wrong, but the two approaches would look
something like this:

Password approach
1. User lands on homepage > 2. user clicks signup > 3. User enters info and
clicks finish > User has access

Auth0 approach
1. User lands on homepage > 2. user clicks signup > 3. User reads message
instructing them to open email > 4. User opens email > 5. User
copies/memorises code > 6. User returns to homepage > 7. User pastes/types
code and clicks finish > 8. User has access

Is this correct?

<https://signup.earwig.work/>

**Comment by Jbarget** [comment]
No problem :) 

Apologies, there are two things here:
1. authentication strategy (standard email + password vs passwordless)
2. who manages the user information (us in our database or auth0 in theirs)

**Comment by thejoefriel** [comment]
Also @theharrisonmoore - in the user stories you have "I am notified of my successful account creation via email (I must click the link in the email to confirm)" so even with a password approach they would have to go to their email to confirm wouldn't they? Unless you want us to remove this email element which would make things a bit simpler.

**Comment by theharrisonmoore** [comment]
Yes, you're correct about there being an email link feature in the User
Stories, and this is there because we have to get 'double opt-in' for GDPR
purposes. Users can register and start using the site without clicking the
link in their email, but later they have to click a link in their email if
they are to confirm they *did* sign up and they are happy to receive
marketing material. If I don't have double opt-in, investors won't touch me
with a barge pole.

Double opt-in wouldn't stop them entering and using the platform, as it
sounds like the Auth0 approach might do.

The bottom line is that we must aim to make the onboarding process a
streamlined as possible. Most people are used to emails and passwords, and
I don't see why we can't keep it simple like this. The passwordless aspect
of Auth0 sounds useful, but I'm not convinced it's going to make onboarding
and repeat visits to the site streamlined, it sounds like it might actually
introduce friction.

What do you think? Am I missing something?
<https://signup.earwig.work/>

**Comment by thejoefriel** [comment]
OK so it would be a standard email and password form - this is also fine for us to set up if preferred. In terms of the double opt-in @theharrisonmoore , do we need a check box on the sign up form as that's not in the wireframe currently?

**Comment by theharrisonmoore** [comment]
With the standard email and password thing, it's still okay to have a 'keep me logged in' tickbox, right? Just trying to reduce login friction - the workers don't have the most patience.

With GDPR, there is a tick box at sign-up where they agree to the earwig Terms. Then they receive an email containing a link and, if and when they click it, we get them on double opt-in.

**Comment by thejoefriel** [comment]
we don't actually need the 'keep me logged in' tickbox. General best practice now anyway is that users want to be kept logged in for a set period. This is absolutely fine as long as you also give them the option to log out.

Re double opt-in, that now makes sense, thanks for explaining. We'll look into this in more detail when we're looking to implement it in the second sprint.

**Comment by theharrisonmoore** [comment]
Fantastic
<https://signup.earwig.work/>
